akka-cluster/src/main/scala/akka/cluster/Cluster.scala

```diff
@@ -359,6 +364,37 @@ class Cluster(val system: ExtendedActorSystem) extends Extension {
   def joinSeedNodes(seedNodes: java.util.List[Address]): Unit =
     joinSeedNodes(Util.immutableSeq(seedNodes))
 
+  /**
+   * Scala API: If the `appVersion` is read from an external system (e.g. Kubernetes) it can be defined after
+   * system startup but before joining by completing the `appVersion` `Future`. In that case, `setAppVersionLater`
+   * should be called before calling `join` or `joinSeedNodes`. It's fine to call `join` or `joinSeedNodes`
+   * immediately afterwards (before the `Future` is completed. The join will then wait for the `appVersion`
+   * to be completed.
+   */
+  def setAppVersionLater(appVersion: Future[Version]): Unit = {
```

Comment:
Since it can only be done once, could it make sense to define it more declaratively through the `akka.actor.setup.Setup` API instead (but the concrete setup living here in cluster ofc)?

Reply:
It will be called from an extension in Akka Management rolling-update-kubernetes. We have extensions for all other things in Akka Managment so having it as a Setup wouldn't align well with existing api.

Also, note that we don't have Setup for typed.ActorSystem.

Reply:
Ah, yeah, I see, setup is fed to actor system as well, so not really possible from extensions.